Core/pinot-dbus-daemon.1
.\" DO NOT MODIFY THIS FILE!  It was generated by help2man 1.40.12.
.TH PINOT-DBUS-DAEMON "1" "February 2013" "pinot-dbus-daemon - pinot 1.04" "User Commands"
.SH NAME
pinot-dbus-daemon \- D-Bus search and index daemon
.SH SYNOPSIS
.B pinot-dbus-daemon
[\fIOPTIONS\fR]
.SH DESCRIPTION
pinot\-dbus\-daemon \- D\-Bus search and index daemon
.SH OPTIONS
.TP
\fB\-h\fR, \fB\-\-help\fR
display this help and exit
.TP
\fB\-i\fR, \fB\-\-ignore\-version\fR
ignore the index version number
.TP
\fB\-p\fR, \fB\-\-priority\fR
set the daemon's priority (default 15)
.TP
\fB\-r\fR, \fB\-\-reindex\fR
force a reindex
.TP
\fB\-v\fR, \fB\-\-version\fR
output version information and exit
.SH "REPORTING BUGS"
Report bugs to fabrice.colin@gmail.com
.PP
.br
This is free software.  You may redistribute copies of it under the terms of
the GNU General Public License <http://www.gnu.org/licenses/old\-licenses/gpl\-2.0.html>.
.br
There is NO WARRANTY, to the extent permitted by law.


NEWS
2013/02/11 version_1_0_4
IndexSearch:
 - fix stripping of diacritics
 - stem subject terms

2013/01/14 version_1_0_3
IndexSearch :
 - fix a Unicode handling issue introduced in 1.01
UI :
 - updated French translation by Eliovir

2012/11/04 version_1_0_2
General :
 - turn memory pooling off by default to avoid issues with newer boost
UI :
 - new Czech translation by Zbyněk Schwarz
 - updated Japanese translation by Takafumi Arakaki
 - updated Brazilian Portuguese translation by Adriano Steffler

2012/08/27 version_1_0_1
General :
 - run rst2html on RST files, if available, and if RST files are detected as such.
 See http://code.google.com/p/pinot-search/issues/detail?id=12
Tokenize :
 - better mbox parts extraction
IndexSearch :
 - dropped unac in favour of own code, resulting in faster indexing.
 - pinot-index --override MIMETYPE:EXTENSION overrides MIME type detection
 based on files extensions

2012/06/16 version_1_0_0
General :
 - install headers and libraries
 - prefer default programs that don't support URIs to view local files
Deskbar :
 - dropped support for Deskbar since it's now dead
Tokenize :
 - better mbox parsing. The internal part numbering scheme has changed;
 reindexing email is recommended.
IndexSearch :
 - support for LibreOffice's libexttextcat v3.2
 - fixed the Google plugin
 - renamed the Freshmeat plugin to Freecode
UI :
 - updated Simplified Chinese translation by happymeng
 - updated German translation by Gena Haltmair
 - initial GTK+ 3 port. Enable with "./configure ... --enable-gtkmm3=yes"

2011/11/07 version_0_9_8
Tokenize :
 - new exiv2-based filter
 - new chmlib-based filter
IndexSearch :
 - support for LibreOffice's libexttextcat v3.1, and possibly v3.1.1
 - dropped plugins for Yahoo! REST API, Yahoo! BOSS, Google Code Search
  and RollYO
UI :
 - query results and view history are expired after 6 months
 - updated Dutch translation by Martijn Verstrate and Tico
 - updated German translation by Fitoschido
 - updated Russian translation by Nikolay Kachanov
 - updated Spanish translation by pkramerruiz

2011/01/09 version_0_9_7
General :
 - replaced custom memory pool class with Boost's
 - don't try and map more than 2Gb and use shared mappings
Deskbar :
 - install the module where DeskBar > 2.28 expects it to be
IndexSearch :
 - remove dots at the end of terms that don't look like acronyms
 - index components of acronyms and dot-separated terms on their own
Daemon :
 - index files as they are crawled, don't delegate indexing to other
  threads, unless PINOT_MAXIMUM_INDEX_THREADS > 1
 - fixed checking of symlinks against black-list
UI :
 - the maximum number of results returned by the Query field and used
  to initialize new stored queries follows PINOT_MAXIMUM_QUERY_RESULTS
 - when a spelling suggestion is available, don't show the same revised
  query multiple times
 - updated Simplified Chinese translation by mike2718
 - updated Dutch translation by Dirk Roos
 - updated Italian translation by Davide Vidal and Simone Sandri
 - updated Japanese translation by Mizuki-san
 - updated Brazilian Portuguese translation by feen
 - updated Portuguese translation by Almufadado
 - updated Russian translation by Alexander Zinin and Nikolay Kachanov
 - updated Spanish translation by Juan Miguel Boyero Corral, Matias Fonzo
  and Fitoschido

2010/07/12 version_0_9_6
General :
 - fixed "GIO can sniff PNG" program used at configure time
 - README clarifies that operators should be upper-case
 - builds with gmime-2.4 or 2.6
 - link with the library that has dlopen() and fix Debian bug #556062
 - merged Debian's patch for --as-needed support
 - merged FreeBSD build patch
 - dropped support for Xesam
SQL :
 - fixed prepared statements interface to work with insertion and deletion
 - sleep then retry operations if the database is busy
 - better transaction support
Tokenize :
 - the mbox filter now supports messages of type "message/external-body"
IndexSearch :
 - fixed possible crash at exit time when the textcat configuration file
  points to non-existing model files
Xapian :
 - prefer the Chert back-end if available. Applicable to Xapian >= 1.2.0.
Daemon :
 - check symlinks against black-list
 - the battery status can now be obtained from DeviceKit-power or upower
UI :
 - updated Simplified Chinese translation by Eleanor Chen
 - updated French translation by verdy_p and Fabrice Colin
 - updated Hebrew translation by Yaron
 - updated Brazilian Portuguese translation by andbelo

2009/11/14 version_0_9_5
General :
 - OpenBSD support, thanks to the work of Antoine Jacoutot
 - fixed build when HAVE_DBUS isn't set
SQL :
 - use prepared statements on most common queries, transactions on mass updates
Tokenize :
 - better handling of acronyms
IndexSearch :
 - updated Bing plugin
 - removed plugins for Exalead and IOI
Xapian :
 - fixed the "path:" operator. Reindexing may be necessary
 - rewrote Search This For feature
UI :
 - if gtkmm >= 2.16 is available, the Find button is replaced with an icon
 - updated Dutch translation by JW
 - updated French translation by Thierry Thomas
 - updated German translation by Fabian Affolter and Marco Jahn
 - updated Hebrew translation by Ddorda
 - updated Portuguese translation by Bernardo Lopes
 - updated Spanish translation by Jesus Tramullas and DiegoJ

2009/06/27 version_0_9_4
General :
 - set _FILE_OFFSET_BITS=64 and fix Debian bug #530572

Core/pinot-dbus-daemon.cc
/*
 *  Copyright 2005-2008 Fabrice Colin
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
 */

#include <stdlib.h>
#include <libintl.h>
#include <getopt.h>
#include <sys/time.h>
#include <sys/resource.h>
#include <sys/types.h>
#include <signal.h>
#include <unistd.h>
#include <errno.h>
#include <iostream>
#include <sstream>
#include <fstream>
#include <cstring>
#include <sigc++/sigc++.h>
#include <glibmm.h>
#include <glibmm/thread.h>
#include <glibmm/ustring.h>
#include <glibmm/miscutils.h>
#include <glibmm/convert.h>

#include "config.h"
#include "NLS.h"
#include "FilterFactory.h"
#include "Languages.h"
#include "MIMEScanner.h"
#include "ModuleFactory.h"
#include "ActionQueue.h"
#include "CrawlHistory.h"
#include "MetaDataBackup.h"
#include "QueryHistory.h"
#include "ViewHistory.h"
#include "DownloaderInterface.h"
#include "DaemonState.h"
#include "PinotSettings.h"
#include "ServerThreads.h"
#include "UniqueApplication.h"

using namespace std;

static ofstream g_outputFile;
static string g_pidFileName;
static streambuf *g_coutBuf = NULL;
static streambuf *g_cerrBuf = NULL;
static struct option g_longOptions[] = {
	{"fullscan", 0, 0, 'f'},
	{"help", 0, 0, 'h'},
	{"ignore-version", 0, 0, 'i'},
	{"priority", 1, 0, 'p'},
	{"reindex", 0, 0, 'r'},
	{"version", 0, 0, 'v'},
	{0, 0, 0, 0}
};
static const char *g_pinotDBusService = "de.berlios.Pinot";
static const char *g_pinotDBusObjectPath = "/de/berlios/Pinot";
static void unregisteredHandler(DBusConnection *pConnection, void *pData);
static DBusHandlerResult messageHandler(DBusConnection *pConnection, DBusMessage *pMessage, void *pData);
static DBusObjectPathVTable g_callVTable = {
	(DBusObjectPathUnregisterFunction)unregisteredHandler,
        (DBusObjectPathMessageFunction)messageHandler,
	NULL,
};
static Glib::RefPtr<Glib::MainLoop> g_refMainLoop;

static void closeAll(void)
{
	cout << "Exiting..." << endl;

	// Close everything
	ModuleFactory::unloadModules();
	Dijon::FilterFactory::unloadFilters();
	Dijon::HtmlFilter::shutdown();

	// Restore the stream buffers
	if (g_coutBuf != NULL)
	{
		cout.rdbuf(g_coutBuf);
	}
	if (g_cerrBuf != NULL)
	{
		cerr.rdbuf(g_cerrBuf);
	}
	g_outputFile.close();
	if (g_pidFileName.empty() == false)
	{
		unlink(g_pidFileName.c_str());
	}

	DownloaderInterface::shutdown();
	MIMEScanner::shutdown();
}

static void quitAll(int sigNum)
{
	if (g_refMainLoop->is_running() == true)
	{
		cout << "Quitting..." << endl;

		g_refMainLoop->quit();
	}
}

static DBusHandlerResult filterHandler(DBusConnection *pConnection, DBusMessage *pMessage, void *pData)
{
	DaemonState *pServer = (DaemonState *)pData;

#ifdef DEBUG
	cout << "filterHandler: called" << endl;
#endif
	// Are we about to be disconnected ?
	if (dbus_message_is_signal(pMessage, DBUS_INTERFACE_LOCAL, "Disconnected") == TRUE)
	{
#ifdef DEBUG
		cout << "filterHandler: received Disconnected" << endl;
#endif
		if (pServer != NULL)
		{
			pServer->set_flag(DaemonState::DISCONNECTED);
		}
		quitAll(0);
	}
	else if (dbus_message_is_signal(pMessage, DBUS_INTERFACE_DBUS, "NameOwnerChanged") == TRUE)
	{
#ifdef DEBUG
		cout << "filterHandler: received NameOwnerChanged" << endl;
#endif
	}
	// The first two signals are specified by the freedesktop.org Power Management spec v0.1 and v0.2
	else if ((dbus_message_is_signal(pMessage, "org.freedesktop.PowerManagement", "BatteryStateChanged") == TRUE) ||
		(dbus_message_is_signal(pMessage, "org.freedesktop.PowerManagement", "OnBatteryChanged") == TRUE) ||
		(dbus_message_is_signal(pMessage, "org.gnome.PowerManager", "OnAcChanged") == TRUE))
	{
		DBusError error;
		gboolean onBattery = FALSE;

#ifdef DEBUG
		cout << "filterHandler: received OnBatteryChanged" << endl;
#endif
		dbus_error_init(&error);
		if ((dbus_message_get_args(pMessage, &error,
			DBUS_TYPE_BOOLEAN, &onBattery,
			DBUS_TYPE_INVALID) == TRUE) &&
			(pData != NULL))
		{
			if (dbus_message_is_signal(pMessage, "org.gnome.PowerManager", "OnAcChanged") == TRUE)
			{
				// This tells us if we are on AC, not on battery
				if (onBattery == TRUE)
				{
					onBattery = FALSE;
				}
				else
				{
					onBattery = TRUE;
				}
			}

			if (onBattery == TRUE)
			{
				// We are now on battery
				if (pServer != NULL)
				{
					pServer->set_flag(DaemonState::ON_BATTERY);
					pServer->stop_crawling();
				}

				cout << "System is now on battery" << endl;
			}
			else
			{
				// Back on-line
				if (pServer != NULL)
				{
					pServer->reset_flag(DaemonState::ON_BATTERY);
					pServer->start_crawling();
				}

				cout << "System is now on AC" << endl;
			}
		}
		dbus_error_free(&error);
	}

	return DBUS_HANDLER_RESULT_NOT_YET_HANDLED;
}

static void unregisteredHandler(DBusConnection *pConnection, void *pData)
{
#ifdef DEBUG
	cout << "unregisteredHandler: called" << endl;
#endif
}

static DBusHandlerResult messageHandler(DBusConnection *pConnection, DBusMessage *pMessage, void *pData)
{
	DaemonState *pServer = (DaemonState *)pData;

	if ((pConnection != NULL) &&
		(pMessage != NULL))
	{
		dbus_connection_ref(pConnection);
		dbus_message_ref(pMessage);

		if (pServer != NULL)
		{
			DBusServletInfo *pInfo = new DBusServletInfo(pConnection, pMessage);

			pServer->start_thread(new DBusServletThread(pServer, pInfo));
		}
	}

	return DBUS_HANDLER_RESULT_HANDLED;
}

static bool getBatteryState(DBusGConnection *pBus, const string &name, const string &path,
	const string &method, gboolean &result)
{
	bool callSuccess = true;

	if ((pBus == NULL) ||
		(name.empty() == true) ||
		(path.empty() == true) ||
		(method.empty() == true))
	{
		return false;
	}

	DBusGProxy *pBusProxy = dbus_g_proxy_new_for_name(pBus, name.c_str(),
		path.c_str(), name.c_str());
	if (pBusProxy == NULL)
	{
		return false;
	}

	GError *pError = NULL;
	if (dbus_g_proxy_call(pBusProxy, method.c_str(), &pError,
		G_TYPE_INVALID,
		G_TYPE_BOOLEAN, &result,
		G_TYPE_INVALID) == FALSE)
	{
		if (pError != NULL)
		{
			cerr << "Couldn't get battery state: " << pError->message << endl;
			g_error_free(pError);
		}

		callSuccess = false;
	}

	g_object_unref(pBusProxy);

	return callSuccess;
}

int main(int argc, char **argv)
{
	int longOptionIndex = 0, priority = 15;
	bool resetHistory = false;
	bool resetLabels = false;
	bool fullScan = false;
	bool reindex = false;
	bool ignoreVersion = false;

	// Look at the options
	int optionChar = getopt_long(argc, argv, "fhip:rv", g_longOptions, &longOptionIndex);
	while (optionChar != -1)
	{
		switch (optionChar)
		{
			case 'f':
				fullScan = true;
				break;
			case 'h':
				// Help
				cout << "pinot-dbus-daemon - D-Bus search and index daemon\n\n"
					<< "Usage: pinot-dbus-daemon [OPTIONS]\n\n"
					<< "Options:\n"
					<< "  -f, --fullscan	force a full scan\n"
					<< "  -h, --help		display this help and exit\n"
					<< "  -i, --ignore-version	ignore the index version number\n"
					<< "  -p, --priority	set the daemon's priority (default 15)\n"
					<< "  -r, --reindex		force a reindex\n"
					<< "  -v, --version		output version information and exit\n"
					<< "\nReport bugs to " << PACKAGE_BUGREPORT << endl;
				return EXIT_SUCCESS;
			case 'i':
				ignoreVersion = true;
				break;
			case 'p':
				if (optarg != NULL)
				{
					int newPriority = atoi(optarg);
					if ((newPriority >= -20) &&
						(newPriority < 20))
					{
						priority = newPriority;
					}
				}
				break;
			case 'r':
				reindex = true;
				break;
			case 'v':
				cout << "pinot-dbus-daemon - " << PACKAGE_STRING << "\n\n" 
					<< "This is free software.  You may redistribute copies of it under the terms of\n"
					<< "the GNU General Public License <http://www.gnu.org/licenses/old-licenses/gpl-2.0.html>.\n"
					<< "There is NO WARRANTY, to the extent permitted by law." << endl;
				return EXIT_SUCCESS;
			default:
				return EXIT_FAILURE;
		}

		// Next option
		optionChar = getopt_long(argc, argv, "fhip:rv", g_longOptions, &longOptionIndex);
	}

#if defined(ENABLE_NLS)
	bindtextdomain(GETTEXT_PACKAGE, PACKAGE_LOCALE_DIR);
	bind_textdomain_codeset(GETTEXT_PACKAGE, "UTF-8");
	textdomain(GETTEXT_PACKAGE);
#endif //ENABLE_NLS

	// Initialize threads support before doing anything else
	if (Glib::thread_supported() == false)
	{
		Glib::thread_init();
	}
	// Initialize the GType and the D-Bus thread system
	g_type_init();
#if DBUS_NUM_VERSION > 1000000
	dbus_threads_init_default();
#endif
	dbus_g_thread_init();

	g_refMainLoop = Glib::MainLoop::create();
	Glib::set_application_name("Pinot DBus Daemon");

	// This should make Xapian use Flint rather than Quartz
	Glib::setenv("XAPIAN_PREFER_FLINT", "1");

	// Make the locale follow the environment variables
	setlocale(LC_ALL, "");
	char *pLocale = setlocale(LC_ALL, NULL);
	if (pLocale != NULL)
	{
		string locale(pLocale);

		if (locale != "C")
		{
			bool appendUTF8 = false;

			string::size_type pos = locale.find_last_of(".");
			if ((pos != string::npos) &&
				((strcasecmp(locale.substr(pos).c_str(), ".utf8") != 0) &&
				(strcasecmp(locale.substr(pos).c_str(), ".utf-8") != 0)))
			{
				locale.resize(pos);
				appendUTF8 = true;
			}

			if (appendUTF8 == true)
			{
				locale += ".UTF-8";

				pLocale = setlocale(LC_ALL, locale.c_str());
				if (pLocale != NULL)
				{
#ifdef DEBUG
					cout << "Changed locale to " << pLocale << endl;
#endif
				}
			}
		}
	}

	// Make sure only one instance runs
	UniqueApplication uniqueApp("de.berlios.PinotDBusDaemon");
	string confDirectory = PinotSettings::getConfigurationDirectory();
	g_pidFileName = confDirectory + "/pinot-dbus-daemon.pid";
	if (chdir(confDirectory.c_str()) == 0)
	{
		if (uniqueApp.isRunning(g_pidFileName, "pinot-dbus-daemon") == true)
		{
			return EXIT_SUCCESS;
		}

		// Redirect cout and cerr to a file
		string fileName(confDirectory);
		fileName += "/pinot-dbus-daemon.log";
		g_outputFile.open(fileName.c_str());
		g_coutBuf = cout.rdbuf();
		g_cerrBuf = cerr.rdbuf();
		cout.rdbuf(g_outputFile.rdbuf());
		cerr.rdbuf(g_outputFile.rdbuf());
	}
	else
	{
		// We can't rely on the PID file
		if (uniqueApp.isRunning() == true)
		{
			return EXIT_SUCCESS;
		}
	}

	// This will create the necessary directories on the first run
	PinotSettings &settings = PinotSettings::getInstance();
	// This is the daemon so disable client-side code 
	settings.enableClientMode(false);

	// Initialize utility classes
	if (MIMEScanner::initialize(PinotSettings::getHomeDirectory() + "/.local",
		string(SHARED_MIME_INFO_PREFIX)) == false)
	{
		cerr << "Couldn't load MIME settings" << endl;
	}
	DownloaderInterface::initialize();
	// Load tokenizer libraries, if any
	Dijon::HtmlFilter::initialize();
	Dijon::FilterFactory::loadFilters(string(LIBDIR) + "/pinot/filters");
	Dijon::FilterFactory::loadFilters(confDirectory + "/filters");
	// Load backends, if any
	ModuleFactory::loadModules(string(LIBDIR) + "/pinot/backends");
	ModuleFactory::loadModules(confDirectory + "/backends");

	// Localize language names
	Languages::setIntlName(0, _("Unknown"));
	Languages::setIntlName(1, _("Danish"));
	Languages::setIntlName(2, _("Dutch"));
	Languages::setIntlName(3, _("English"));
	Languages::setIntlName(4, _("Finnish"));
	Languages::setIntlName(5, _("French"));
	Languages::setIntlName(6, _("German"));
	Languages::setIntlName(7, _("Hungarian"));
	Languages::setIntlName(8, _("Italian"));
	Languages::setIntlName(9, _("Norwegian"));
	Languages::setIntlName(10, _("Portuguese"));
	Languages::setIntlName(11, _("Romanian"));
	Languages::setIntlName(12, _("Russian"));
	Languages::setIntlName(13, _("Spanish"));
	Languages::setIntlName(14, _("Swedish"));
	Languages::setIntlName(15, _("Turkish"));

	// Load the settings
	settings.load(PinotSettings::LOAD_ALL);

	// Catch interrupts
#ifdef HAVE_SIGACTION
	struct sigaction newAction;
	sigemptyset(&newAction.sa_mask);
	newAction.sa_flags = 0;
	newAction.sa_handler = quitAll;
	sigaction(SIGINT, &newAction, NULL);
	sigaction(SIGQUIT, &newAction, NULL);
	sigaction(SIGTERM, &newAction, NULL);
#else
	signal(SIGINT, quitAll);
#ifdef SIGQUIT
	signal(SIGQUIT, quitAll);
#endif
	signal(SIGTERM, quitAll);
#endif

	// Open the daemon index in read-write mode 
	bool wasObsoleteFormat = false;
	if (ModuleFactory::openOrCreateIndex(settings.m_defaultBackend, settings.m_daemonIndexLocation, wasObsoleteFormat, false) == false)
	{
		cerr << "Couldn't open index " << settings.m_daemonIndexLocation << endl;
		return EXIT_FAILURE;
	}
	if (wasObsoleteFormat == true)
	{
		resetHistory = resetLabels = true;
	}

	// Do the same for the history database
	PinotSettings::checkHistoryDatabase();
	string historyDatabase(settings.getHistoryDatabaseName());
	if ((historyDatabase.empty() == true) ||
		(ActionQueue::create(historyDatabase) == false) ||
		(CrawlHistory::create(historyDatabase) == false) ||
		(MetaDataBackup::create(historyDatabase) == false) ||
		(QueryHistory::create(historyDatabase) == false) ||
		(ViewHistory::create(historyDatabase) == false))
	{
		cerr << "Couldn't create history database " << historyDatabase << endl;
		return EXIT_FAILURE;
	}
	else
	{
		ActionQueue actionQueue(historyDatabase, Glib::get_application_name());
		QueryHistory queryHistory(historyDatabase);
		ViewHistory viewHistory(historyDatabase);
		time_t timeNow = time(NULL);
		unsigned int actionsCount = actionQueue.getItemsCount(ActionQueue::INDEX);

		// Don't expire actions left from last time
		actionsCount += actionQueue.getItemsCount(ActionQueue::UNINDEX);
		cout << actionsCount << " actions left" << endl;

		// Expire the rest
		queryHistory.expireItems(timeNow);
		viewHistory.expireItems(timeNow);
	}

	atexit(closeAll);

	// Change the daemon's priority
	if (setpriority(PRIO_PROCESS, 0, priority) == -1)
	{
		cerr << "Couldn't set scheduling priority to " << priority << endl;
	}
#ifdef DEBUG
	else cout << "Set priority to " << priority << endl;
#endif

	GError *pError = NULL;
	DBusGConnection *pBus = dbus_g_bus_get(DBUS_BUS_SESSION, &pError);
	if (pBus == NULL)
	{
		if (pError != NULL)
		{
			cerr << "Couldn't open bus connection: " << pError->message << endl;
			if (pError->message != NULL)
			{
				cerr << "Error is " << pError->message << endl;
			}
			g_error_free(pError);
		}

		return EXIT_FAILURE;
	}

	DBusConnection *pConnection = dbus_g_connection_get_connection(pBus);
	if (pConnection == NULL)
	{
		cerr << "Couldn't get connection" << endl;
		return EXIT_FAILURE;
	}

	DBusError error;
	DaemonState server;
	IndexInterface *pIndex = NULL;

	dbus_error_init(&error);
	dbus_connection_set_exit_on_disconnect(pConnection, FALSE);
	dbus_connection_setup_with_g_main(pConnection, NULL);

	if (dbus_connection_register_object_path(pConnection, g_pinotDBusObjectPath,
		&g_callVTable, &server) == TRUE)
	{
		// Request to be identified by this name
		// FIXME: flags are currently broken ?
		dbus_bus_request_name(pConnection, g_pinotDBusService, 0, &error);
		if (dbus_error_is_set(&error) == FALSE)
		{
			// See power management signals
			dbus_bus_add_match(pConnection,
				"type='signal',interface='org.freedesktop.PowerManagement'", &error);
			dbus_bus_add_match(pConnection,
				"type='signal',interface='org.gnome.PowerManager'", &error);

			dbus_connection_add_filter(pConnection,
				(DBusHandleMessageFunction)filterHandler, &server, NULL);
		}
		else
		{
			cerr << "Couldn't obtain name " << g_pinotDBusService << endl;
			if (error.message != NULL)
			{
				cerr << "Error is " << error.message << endl;
			}
		}

		try
		{
			set<string> labels;
			bool gotLabels = false;
			bool onBattery = false;

			pIndex = settings.getIndex(settings.m_daemonIndexLocation);
			if (pIndex != NULL)
			{
				string indexVersion(pIndex->getMetadata("version"));

				gotLabels = pIndex->getLabels(labels);

				// What version is the index at ?
				if (indexVersion.empty() == true)
				{
					indexVersion = "0.0";
				}
				if (ignoreVersion == true)
				{
					// Better reset labels, they may have been lost too
					resetLabels = true;
				}
				else if (indexVersion < PINOT_INDEX_MIN_VERSION)
				{
					cout << "Upgrading index from version " << indexVersion << " to " << VERSION << endl;

					reindex = true;
				}
				if (reindex == true)
				{
					// Reset the index so that all documents are reindexed
					pIndex->reset();

					cout << "Reset index" << endl;

					resetHistory = resetLabels = true;
				}

				pIndex->setMetadata("version", VERSION);
				pIndex->setMetadata("dbus-status", "Running");
			}

			if (resetHistory == true)
			{
				CrawlHistory history(historyDatabase);
				map<unsigned int, string> sources;

				// Reset the history
				history.getSources(sources);
				for (std::map<unsigned int, string>::iterator sourceIter = sources.begin();
					sourceIter != sources.end(); ++sourceIter)
				{
					history.deleteItems(sourceIter->first);
					history.deleteSource(sourceIter->first);
				}

				cout << "Reset crawler history" << endl;
			}

			if ((resetLabels == true) &&
				(pIndex != NULL))
			{
				// Re-apply the labels list
				if (gotLabels == false)
				{
					// If this is an upgrade from a version < 0.80, the labels list
					// needs to be pulled from the configuration file
					pIndex->setLabels(settings.m_labels, true);

					cout << "Set labels as per the configuration file" << endl;
				}
				else
				{
					pIndex->setLabels(labels, true);
				}
			}

			// Connect to the quit signal
			server.getQuitSignal().connect(sigc::ptr_fun(&quitAll));

			// Connect to threads' finished signal
			server.connect();

			// Try and get the battery state
			gboolean result = FALSE;
			if ((getBatteryState(pBus, "org.freedesktop.PowerManagement",
				"/org/freedesktop/PowerManagement", "GetOnBattery", result) == true) ||
				(getBatteryState(pBus, "org.freedesktop.PowerManagement",
				"/org/freedesktop/PowerManagement", "GetBatteryState", result) == true))
			{
				if (result == TRUE)
				{
					onBattery = true;
				}
			}
			else if (getBatteryState(pBus, "org.gnome.PowerManager",
				"/org/gnome/PowerManager", "GetOnAc", result) == true)
			{
				if (result == FALSE)
				{
					onBattery = true;
				}
			}
			if (onBattery == true)
			{
				// We are on battery
				server.set_flag(DaemonState::ON_BATTERY);
				server.stop_crawling();

				cout << "System is on battery" << endl;
			}

			server.start(fullScan, reindex);

			// Run the main loop
			g_refMainLoop->run();

		}
		catch (const Glib::Exception &e)
		{
			cerr << e.what() << endl;
			return EXIT_FAILURE;
		}
		catch (const char *pMsg)
		{
			cerr << pMsg << endl;
			return EXIT_FAILURE;
		}
		catch (...)
		{
			cerr << "Unknown exception" << endl;
			return EXIT_FAILURE;
		}
	}
	else
	{
		cerr << "Couldn't register object path" << endl;
	}
	dbus_error_free(&error);

	if (pIndex != NULL)
	{
		if (server.is_flag_set(DaemonState::DISCONNECTED) == true)
		{
			pIndex->setMetadata("dbus-status", "Disconnected");
		}
		else if (server.is_flag_set(DaemonState::STOPPED) == true)
		{
			pIndex->setMetadata("dbus-status", "Stopped");
		}
		else
		{
			pIndex->setMetadata("dbus-status", "");
		}
		delete pIndex;
	}

	// Stop everything
	server.disconnect();
	server.stop_threads();

	return EXIT_SUCCESS;
}
